src/librustc_mir/build/matches/mod.rs

```diff
@@ -1790,3 +1762,68 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         self.var_indices.insert(var_id, locals);
     }
 }
+
+// HACK(matthewjasper | dlrobertson) This is a trait so that we
+// can let the test create its blocks before the rest of the match.
+// This currently improves the speed of llvm when optimizing long
+// string literal matches
+trait TargetBlockBuilder<'tcx> {
+    fn target_blocks(self, this: &mut Builder<'_, 'tcx>) -> Vec<BasicBlock>;
+}
+
+struct CandidateTargetBlockBuilder<'tcx, 'pat, 'a, 'b, 'c> {
+    candidates: &'a mut [&'b mut Candidate<'pat, 'tcx>],
+    span: Span,
+    fake_borrows: &'c mut Option<FxHashSet<Place<'tcx>>>,
+    otherwise_block: Option<BasicBlock>,
+    target_candidates: Vec<Vec<&'c mut Candidate<'pat, 'tcx>>>,
+}
+
+impl<'tcx, 'pat, 'a, 'b, 'c> TargetBlockBuilder<'tcx> for
+        CandidateTargetBlockBuilder<'tcx, 'pat, 'a, 'b, 'c> {
+    fn target_blocks(mut self, this: &mut Builder<'_, 'tcx>) -> Vec<BasicBlock> {
+        // For each outcome of test, process the candidates that still
+        // apply. Collect a list of blocks where control flow will
+        // branch if one of the `target_candidate` sets is not
+        // exhaustive.
+        if !self.candidates.is_empty() {
+            let remainder_start = &mut None;
+            this.match_candidates(
+                self.span,
+                remainder_start,
+                self.otherwise_block,
+                self.candidates,
+                self.fake_borrows,
+            );
+            self.otherwise_block = Some(remainder_start.unwrap());
+        };
+
+        let mut otherwise_block = self.otherwise_block;
+        let span = self.span;
+        let fake_borrows = self.fake_borrows;
+        self.target_candidates.into_iter().map(|mut candidates| {
+            if candidates.len() != 0 {
```

Comment:
`!candidates.is_empty()` is more canonical.

Reply:
Good point